pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -133,35 +137,68 @@ func (r *Reconciler) deleteClusterResourceSnapshots(ctx context.Context, crp *fl
 			// use the default value instead
 			revisionLimit = fleetv1beta1.RevisionHistoryLimitDefaultValue
 		}
 	}
 
-	_, err := r.getOrCreateClusterSchedulingPolicySnapshot(ctx, crp, int(revisionLimit))
+	latestSchedulingPolicySnapshot, err := r.getOrCreateClusterSchedulingPolicySnapshot(ctx, crp, int(revisionLimit))
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	selectedResources, _, err := r.selectResourcesForPlacement(crp)
+	selectedResources, selectedResourceIDs, err := r.selectResourcesForPlacement(crp)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 	resourceSnapshotSpec := fleetv1beta1.ResourceSnapshotSpec{
 		SelectedResources: selectedResources,
 	}
-	_, err = r.getOrCreateClusterResourceSnapshot(ctx, crp, &resourceSnapshotSpec, int(revisionLimit))
+	latestResourceSnapshot, err := r.getOrCreateClusterResourceSnapshot(ctx, crp, &resourceSnapshotSpec, int(revisionLimit))
 	if err != nil {
 		return ctrl.Result{}, err
 	}
+	if err := r.setPlacementStatus(ctx, crp, selectedResourceIDs, latestSchedulingPolicySnapshot, latestResourceSnapshot); err != nil {
+		return ctrl.Result{}, err
+	}
 
-	// update the status based on the latestPolicySnapshot status
-	// update the status based on the work
-	return ctrl.Result{}, nil
+	if err := r.Client.Status().Update(ctx, crp); err != nil {
+		klog.ErrorS(err, "Failed to update the status", "clusterResourcePlacement", crpKObj)
+		return ctrl.Result{}, err
+	}
+	klog.V(2).InfoS("Updated the clusterResourcePlacement status", "clusterResourcePlacement", crpKObj)
+
+	if isRolloutCompleted(crp) {
+		if !isRolloutCompleted(oldCRP) {
+			klog.V(2).InfoS("Placement rollout has finished", "clusterResourcePlacement", crpKObj, "generation", crp.Generation)
+			r.Recorder.Event(crp, corev1.EventTypeNormal, "PlacementRolloutCompleted", "Resources have been applied to the selected clusters")
+		}
+		// We keep a slow reconcile loop here to periodically update the work status in case the applied works change the status.
+		klog.V(2).InfoS("Placement rollout has finished and requeue the request in case of works change", "clusterResourcePlacement", crpKObj)
+		return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+	}
+
+	if !isCRPScheduled(oldCRP) && isCRPScheduled(crp) {
+		klog.V(2).InfoS("Placement has been scheduled", "clusterResourcePlacement", crpKObj, "generation", crp.Generation)
+		r.Recorder.Event(crp, corev1.EventTypeNormal, "PlacementScheduleSuccess", "Successfully scheduled the placement")
+	}
+	if !isCRPSynchronized(oldCRP) && isCRPSynchronized(crp) {
+		klog.V(2).InfoS("Placement has been synchronized", "clusterResourcePlacement", crpKObj, "generation", crp.Generation)
+		r.Recorder.Event(crp, corev1.EventTypeNormal, "PlacementSyncSuccess", "Successfully synchronized the placement")
+	}
+	if !isCRPApplied(oldCRP) && isCRPApplied(crp) {
+		klog.V(2).InfoS("Placement has been applied", "clusterResourcePlacement", crpKObj, "generation", crp.Generation)
```

Comment:
can  !isCRPApplied(oldCRP) && isCRPApplied(crp) be "true" but isRolloutCompleted(crp)  false?

which means isCRPApplied(crp)  is true but somehow isCRPScheduled(crp) or  isCRPSynchronized(crp) is false. That's kinda counter intuitive to me.

Reply:
based on the current implementation, it's impossible. will remove it and add a comment instead.